[interp] fix ftnptr_eh_callback

lewurm opened the pull request:
please look at the single commits. here's a `git log` copy/paste:

---------------------------------

[exceptions] ftnptr_eh_callback test that stresses n2m transitions.

on Xamarin.iOS it happens that the ftnptr_eh_callback calls into managed code, that is yet another native-to-managed transition.


----------------------------------


[exceptions] pop lmf stack before executing ftnptr_eh_callback


----------------------------------

[interp] always set LMF for interp_in wrapper.

partial revert of a33cbb873c487490f2a3c271b24c70c1c3fc0a2c



----------------------------------

[mini] ignore LMF if it isn't initialized

interp_in wrapper serves two purposes:

1) native to interp transition: usually this transition doesn't need a
    LMF marker, but in a ftnptr_eh_callback scenario it can be required.
    see ./mono/tests/install_eh_callback.cs
2) jitted code to interp. LMF marker is used to signal EH the transition.

The change of this commit is needed if we have a wrapper of kind 1), and it's called from native code by a thread that isn't attached to the runtime yet. Therefore LMF won't be initialized yet.

BrzVlad commented:
@lewurm The interp lmf_wrapper was added so you can unwind directly into the calling code (with the register state at the site), so it's main purpose is to save the register context. I don't see how this is used for the same purpose here. I'm guessing it just adds a dummy lmf structure that is for whatever reason expected somewhere for the EH to continue unwinding ?

vargaz commented on mono/mini/method-to-ir.c:
```diff
@@ -1740,11 +1740,14 @@ emit_push_lmf (MonoCompile *cfg)
-	 * lmf->prev_lmf = *lmf_addr
-	 * *lmf_addr = lmf
+	 * if (lmf_addr != NULL) {
+	 *   lmf->lmf_addr = lmf_addr
```
Why is this needed ? The lmf should only be null if the thread is not attached to the JIT.

lewurm replied:
the `interp_in` wrapper is called _before_ the `native-to-managed` wrapper which would to do the attaching. if it's called from an unattached thread, we don't care about the LMF in this wrapper; in other scenarios we might do.

vargaz replied:
Not sure why its called before, but if that really happens, then the null check should be emitted only in interp_in wrappers, not everywhere.

BrzVlad commented on mono/mini/interp/interp.c:
```diff
@@ -334,6 +334,28 @@ mono_interp_get_imethod (MonoDomain *domain, MonoMethod *method, MonoError *erro
+	ext->interp_exit_data = frame;
+
+#ifdef MONO_INIT_CONTEXT_FROM_CURRENT
+	MONO_INIT_CONTEXT_FROM_CURRENT (&ext->ctx);
```
@lewurm I don't understand the point of this define path

lewurm replied:
it's copy/paste from other places that are using `MONO_INIT_CONTEXT_FROM_FUNC`. we could probably remove everything that mentions `MONO_INIT_CONTEXT_FROM_CURRENT` from the code base, no architecture seems to define it.

lewurm commented:
My previous attempt was wrong: Previously, I tried to sneak in a `MonoContext` via an LMF wrapper on the native2interp transition. That worked by chance for this testcase, but is wrong and broke the debugger. As @BrzVlad pointed out to me on Slack, we should rather inject `MonoContext` on the interp2native transition which is what the updated PR does. We could have done it with the LMF wrapper as well, but due to `MINT_CALLI_NAT_FAST` (a fast path optimization for common signatures of icalls/pinvokes) we would need to handle a couple different signatures. I tried to avoid that with my solution.

The testcase is unchanged, so I keep @lambdageek's review for that. @BrzVlad @vargaz please have a look at the native bits.

BrzVlad commented on mono/mini/interp/interp.c:
```diff
@@ -334,6 +334,33 @@ mono_interp_get_imethod (MonoDomain *domain, MonoMethod *method, MonoError *erro
+#define INTERP_PUSH_LMF_WITH_CTX_BODY(ext, exit_addr) \
+	MONO_CONTEXT_GET_CURRENT ((ext).ctx); \
+	MONO_CONTEXT_SET_IP (&(ext).ctx, (exit_addr)); \
+	mono_arch_do_ip_adjustment (&(ext).ctx);
```
@lewurm I'm assuming the only purpuse of ip adjustment is for EH to undo it, right ?

lewurm replied:
correct

lambdageek commented:
Discussed with @lewurm - the coop failures are related to the new test but not to the code changes.  We should take this PR and fix the coop breakage separately. (it's a checked build failure - the cookies on the stack got out of sync)

lewurm commented:
removed the test case from this PR and moved it to https://github.com/mono/mono/pull/10337

Here are some CI runs where the interpreter changes worked for mentioned test case:
* https://jenkins.mono-project.com/job/test-mono-pull-request-interpreter/9285/
* https://jenkins.mono-project.com/job/test-mono-pull-request-interpreter-aarch64/7751/
* https://jenkins.mono-project.com/job/test-mono-pull-request-interpreter-armhf/7465/
* https://jenkins.mono-project.com/job/test-mono-pull-request-wasm/4228/
* https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx-products-sdks-ios/4247/

Failing coop jobs:
* https://jenkins.mono-project.com/job/test-mono-pull-request-coop-arm64/3592/
* https://jenkins.mono-project.com/job/test-mono-pull-request-coop/7310/

lewurm commented:
@monojenkins rebase

monojenkins commented:
Cannot rebase because the following required status checks are not successful:
 - "Linux x64 Checked Private Types Build" state is "failure"

lewurm commented:
@monojenkins backport 2018-06

lewurm commented:
@monojenkins backport 2018-08

monojenkins commented:
@lewurm backporting to 2018-06 failed, the patch results in conflicts:

```
Applying: [exceptions] pop lmf stack before executing ftnptr_eh_callback
Applying: [interp] attach LMF with MonoContext in interp2native transition
Using index info to reconstruct a base tree...
M	mono/metadata/loader.c
M	mono/mini/debugger-agent.c
M	mono/mini/debugger-engine.c
M	mono/mini/exceptions-amd64.c
M	mono/mini/exceptions-arm64.c
M	mono/mini/interp/interp.c
M	mono/mini/mini-exceptions.c
M	mono/mini/mini-runtime.h
M	mono/mini/mini-wasm-debugger.c
M	mono/mini/mini.h
Falling back to patching base and 3-way merge...
Auto-merging mono/mini/mini.h
Auto-merging mono/mini/mini-wasm-debugger.c
CONFLICT (content): Merge conflict in mono/mini/mini-wasm-debugger.c
Auto-merging mono/mini/mini-runtime.h
Auto-merging mono/mini/mini-exceptions.c
Auto-merging mono/mini/interp/interp.c
Auto-merging mono/mini/exceptions-arm64.c
Auto-merging mono/mini/exceptions-amd64.c
Auto-merging mono/mini/debugger-engine.c
CONFLICT (content): Merge conflict in mono/mini/debugger-engine.c
Auto-merging mono/mini/debugger-agent.c
CONFLICT (content): Merge conflict in mono/mini/debugger-agent.c
Auto-merging mono/metadata/loader.c
error: Failed to merge in the changes.
Patch failed at 0002 [interp] attach LMF with MonoContext in interp2native transition
```

Please backport manually!
